sdk/dotnet/Docker.cs

```diff
@@ -0,0 +1,644 @@
+                throw new ResourceException($"{GetFailureMessage(cmd, args, reportFullCommandLine, code)}\n{stdout}", logResource);
+            }
+
+            return stdout;
+        }
+
+        /// <summary>
+        /// Runs a CLI command in a child process, returning a promise for the process's exit. Both stdout
+        /// and stderr are redirected to process.stdout and process.stder by default.
+        /// </summary>
+        /// <param name="cmd">Command name to run.</param>
+        /// <param name="args">An array of command arguments.</param>
+        /// <param name="logResource">Used to specify the resource to associate command output with. Stderr messages
+        /// are always sent (since they may contain important information about something that's gone wrong).
+        /// Stdout messages will be logged ephemerally to this resource.  This lets the user know there is
+        /// progress, without having that dumped on them at the end.  If an error occurs though, the stdout
+        /// content will be printed.</param>
+        /// <param name="reportFullCommandLine"></param>
+        /// <param name="reportErrorAsWarning"></param>
+        /// <param name="stdin">If defined, its contents are piped into stdin for the child process.</param>
+        /// <param name="env"></param>
+        /// <returns></returns>
+        private static async Task<(int, string)> RunCommandThatCanFail(
+            string cmd,
+            string[] args,
+            Resource logResource,
+            bool reportFullCommandLine,
+            bool reportErrorAsWarning,
+            string? stdin = null,
+            Dictionary<string, string>? env = null)
+        {
+            // Let the user ephemerally know the command we're going to execute.
+            Log.Info($"Executing {GetCommandLineMessage(cmd, args, reportFullCommandLine, env)}", logResource, ephemeral: true);
+
+            // Generate a unique stream-ID that we'll associate all the docker output with. This will allow
+            // each spawned CLI command's output to associated with 'resource' and also streamed to the UI
+            // in pieces so that it can be displayed live.  The stream-ID is so that the UI knows these
+            // messages are all related and should be considered as one large message (just one that was
+            // sent over in chunks).
+            //
+            // We use Math.random here in case our package is loaded multiple times in memory (i.e. because
+            // different downstream dependencies depend on different versions of us).  By being random we
+            // effectively make it completely unlikely that any two cli outputs could map to the same stream
+            // id.
+            //
+            // Pick a reasonably distributed number between 0 and 2^30.  This will fit as an int32
+            // which the grpc layer needs.
+            var streamID = (int)Math.Floor(new Random().NextDouble() * (1 << 30));
```

Comment:
put `new Random()` in a static, and lock around it.  i don't believe Random is threadsafe, and it shouldn't be recreated each time.

Reply:
Actually, `streamID` is not needed in current .NET implementation

Reply:
Re-introduced `streamId` and a static random generator for async stream of outputs